src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -378,22 +244,20 @@ private bool Steal(out T result, bool take)
 #endif
 
             bool loop;
-            List<int> versionsList = new List<int>(); // save the lists version
+            List<int> versionsList = t_versionsList ?? (t_versionsList = new List<int>());
             do
             {
                 versionsList.Clear(); //clear the list from the previous iteration
                 loop = false;
 
-
-                ThreadLocalList currentList = _headList;
-                while (currentList != null)
+                ThreadLocalList currentList;
+                for (currentList = _headList; currentList != null; currentList = currentList._nextList)
                 {
                     versionsList.Add(currentList._version);
-                    if (currentList._head != null && TrySteal(currentList, out result, take))
+                    if (currentList.TrySteal(out result, take))
                     {
                         return true;
                     }
-                    currentList = currentList._nextList;
                 }
 
                 // verify versioning, if other items are added to this list since we last visit it, we should retry
```

Comment:
I've been trying to come up with a scenario where the versioning helps prevent a "wrong" answer, i.e. where a TryTake is issued after an item has been added but where it's unable to steal that item and where versioning would avoid that... and I've not come up with one.  This does seem superfluous.  I'll remove it if I can't come up with a good reason for it.